File: packages/generator-go/render.go
package generator

import (
	"fmt"
	"regexp"
	"strconv"
	"strings"
)

func escape(s string) string {
	return strings.NewReplacer("&", "&amp;", "<", "&lt;", ">", "&gt;", `"`, "&quot;", "'", "&#x27;").Replace(s)
}
func rawEscape(s string) string {
	return strings.NewReplacer("&", "&amp;", `"`, "&quot;", "<", "&lt;").Replace(s)
}
func escapeText(s string) string {
	return strings.NewReplacer("&", "&amp;", "<", "&lt;", ">", "&gt;").Replace(s)
}
func eventAttrs(a *Object) bool {
	if a == nil {
		return false
	}
	for _, k := range a.Keys() {
		if strings.HasPrefix(k, "on") {
			return true
		}
	}
	return false
}
func attrs(a *Object, raw, input bool) string {
	if a == nil {
		return ""
	}
	out := ""
	value, hasValue := a.Get("value")
	checked, hasChecked := a.Get("checked")
	name, hasName := a.Get("name")
	style, hasStyle := a.Get("style")
	for _, k := range a.Keys() {
		v := read(a, k)
		if isAbsent(v) || v == nil {
			continue
		}
		if !raw && input && (k == "value" || k == "checked" || k == "name" || k == "style") {
			continue
		}
		s := scalar(v)
		if !raw && (k == "href" || k == "src") {
			s = sanitizeURL(s)
		}
		if k == "style" && !raw {
			s = compactStyle(s)
			if s == "" {
				continue
			}
		}
		if !raw && (k == "readonly" || k == "disabled" || k == "required" || k == "multiple" || k == "autofocus") {
			s = ""
		}
		esc := escape
		if raw {
			esc = rawEscape
		}
		key := k
		if !raw {
			key = ordinaryAttribute(k)
		}
		out += " " + key + `="` + esc(s) + `"`
	}
	if !raw && input {
		if hasStyle {
			if css := compactStyle(scalar(style)); css != "" {
				out += ` style="` + escape(css) + `"`
			}
		}
		if hasName {
			out += ` name="` + escape(scalar(name)) + `"`
		}
		if hasChecked && truthy(checked) {
			out += ` checked=""`
		}
		if hasValue {
			out += ` value="` + escape(scalar(value)) + `"`
		}
	}
	return out
}
func element(tag string, a *Object, body string) string {
	return "<" + tag + attrs(a, false, false) + ">" + body + "</" + tag + ">"
}
func inputHTML(a *Object, raw bool) string {
	end := "/>"
	if raw {
		end = ">"
	}
	return "<input" + attrs(a, raw, true) + end
}
func affixHTML(v any, raw bool) string {
	a := object(v)
	if a == nil {
		return ""
	}
	at := NewObject()
	for _, k := range []string{"class", "style"} {
		if s := stringAt(a, k); s != "" {
			at.Set(k, s)
		}
	}
	return "<span" + attrs(at, raw, false) + ">" + textContent(stringAt(a, "text"), raw) + "</span>"
}
func objectList(v any) []*Object {
	if a, ok := v.([]*Object); ok {
		return a
	}
	out := []*Object{}
	for _, a := range list(v) {
		if o := object(a); o != nil {
			out = append(out, o)
		}
	}
	return out
}
func controlHTML(w *Object, raw bool, selection string) string {
	tag := stringAt(w, "tag")
	a := object(read(w, "attrs"))
	if !raw && a.Has("style") {
		a = a.Clone()
		style := read(a, "style")
		a.Delete("style")
		a.Set("style", style)
	}
	switch tag {
	case "select":
		body := ""
		for _, o := range objectList(read(w, "options")) {
			at := NewObject("value", stringAt(o, "value"))
			if truthy(read(o, "selected")) {
				at.Set("selected", selection)
			}
			body += "<option" + attrs(at, raw, false) + ">" + textContent(stringAt(o, "label"), raw) + "</option>"
		}
		return "<select" + attrs(a, raw, false) + ">" + body + "</select>"
	case "textarea":
		text := stringAt(w, "text")
		if !raw && strings.HasPrefix(text, "\n") {
			text = "\n" + text
		}
		return "<textarea" + attrs(a, raw, false) + ">" + textContent(text, raw) + "</textarea>"
	default:
		return inputHTML(a, raw)
	}
}
func renderWidget(w *Object) string {
	if w == nil {
		return ""
	}
	if read(w, "unsupported") == true {
		return element("div", NewObject("class", "form-element-unsupported", "data-unsupported-type", stringAt(w, "type")), "")
	}
	a := object(read(w, "attrs"))
	raw := eventAttrs(a)
	script := `<script nonce="">` + stringAt(w, "script") + `</script>`
	switch stringAt(w, "layout") {
	case "input-group":
		return `<div class="input-group">` + affixHTML(read(w, "prepend"), raw) + controlHTML(w, raw, "") + affixHTML(read(w, "append"), raw) + `</div>`
	case "bare":
		return controlHTML(w, raw, "")
	case "host-script":
		return controlHTML(w, raw, "") + script
	case "btn-group":
		shared := object(read(read(w, "extra"), "input"))
		raw = eventAttrs(shared)
		body := ""
		typ := "checkbox"
		if stringAt(w, "kind") == "choice" {
			typ = "radio"
		}
		for _, o := range objectList(read(w, "options")) {
			at := NewObject()
			merge(at, shared)
			at.Set("type", typ)
			at.Set("value", stringAt(o, "value"))
			at.Set("autocomplete", "off")
			at.Set("class", "valid-target btn-check")
			if id := stringAt(o, "id"); id != "" {
				at.Set("id", id)
			}
			if typ == "radio" {
				v := ""
				if truthy(read(o, "isDefault")) {
					v = "1"
				}
				at.Set("data-is-default", v)
			}
			if truthy(read(o, "selected")) {
				at.Set("checked", true)
			}
			if raw && at.Has("checked") {
				at.Set("checked", "")
			}
			body += inputHTML(at, raw)
			label := NewObject()
			if id := stringAt(o, "id"); id != "" {
				label.Set("for", id)
			}
			label.Set("class", stringAt(w, "itemLabelClass"))
			body += "<label" + attrs(label, raw, false) + "><span>" + textContent(stringAt(o, "label"), raw) + "</span></label>"
		}
		return "<div" + attrs(a, false, false) + ">" + body + "</div>"
	case "file":
		extra := read(w, "extra")
		file := object(read(extra, "file"))
		display := object(read(extra, "display"))
		raw = eventAttrs(file)
		body := affixHTML(read(w, "prepend"), raw)
		if display != nil {
			if raw {
				body += `<input class="` + rawEscape(stringAt(display, "class")) + `" readonly="" type="text" value="">`
			} else {
				body += inputHTML(display, false)
			}
		}
		body += inputHTML(file, raw)
		if display != nil {
			body += `<button class="btn btn-search btn-file-search" type="button">&nbsp;</button>`
		}
		return `<div class="input-group">` + body + `</div>`
	case "display":
		return "<div" + attrs(a, false, false) + ">" + stringAt(w, "rawHtml") + "</div>"
	case "search":
		style := ""
		if s := stringAt(w, "styleChrome"); s != "" {
			style = `<style nonce="">` + s + `</style>`
		}
		return style + script + `<div class="input-group field-search">` + affixHTML(read(w, "prepend"), true) + controlHTML(w, true, "selected") + affixHTML(read(w, "append"), true) + `</div>`
	case "button":
		return script + inputHTML(object(read(read(w, "extra"), "hidden")), false) + inputHTML(a, false)
	}
	return ""
}

// classes joins the nonempty class parts with single spaces.
func classes(parts ...string) string {
	out := []string{}
	for _, p := range parts {
		if p != "" {
			out = append(out, p)
		}
	}
	return strings.Join(out, " ")
}

// flag renders a valueless boolean attribute.
func flag(name string, on bool) string {
	if on {
		return " " + name + `=""`
	}
	return ""
}
func controlsHTML(controls *Object) string {
	body := ""
	for _, a := range objectList(read(controls, "actions")) {
		body += "<button" + attrs(NewObject("type", "button", "class", "crudui-action", "data-crudui-action", stringAt(a, "name"), "aria-label", stringAt(a, "label")), false, false) + flag("disabled", read(a, "disabled") == true) + "></button>"
	}
	return element("div", NewObject("class", "crudui-controls", "role", "group", "aria-label", stringAt(controls, "label")), body)
}
func headerHTML(vm *Object) string {
	header := read(vm, "header")
	expanded := read(vm, "expanded") == true
	parts := ""
	if read(vm, "collapsible") == true {
		a := NewObject("type", "button", "class", "crudui-action", "data-crudui-action", "toggle-row", "aria-expanded", strconv.FormatBool(expanded))
		if body := read(vm, "body"); has(body, "id") {
			a.Set("aria-controls", stringAt(body, "id"))
		}
		if has(vm, "toggleLabel") {
			a.Set("aria-label", stringAt(vm, "toggleLabel"))
		}
		parts += "<button" + attrs(a, false, false) + "></button>"
	}
	if has(header, "label") {
		if target := stringAt(header, "labelFor"); target != "" {
			parts += element("label", NewObject("class", "crudui-node__label", "for", target), escape(stringAt(header, "label")))
		} else {
			parts += element("span", NewObject("class", "crudui-node__label"), escape(stringAt(header, "label")))
		}
	}
	if has(header, "description") {
		parts += element("p", NewObject("class", "crudui-node__description"), escape(stringAt(header, "description")))
	}
	for _, part := range []string{"number", "title"} {
		if has(header, part) {
			parts += element("span", NewObject("class", "crudui-node__"+part), escape(stringAt(header, part)))
		}
	}
	if has(header, "summary") {
		parts += `<span class="crudui-node__summary"` + flag("hidden", expanded) + ">" + escape(stringAt(header, "summary")) + "</span>"
	}
	if has(header, "count") {
		parts += element("span", NewObject("class", "crudui-node__count"), escape(stringAt(header, "count")))
	}
	controls := read(vm, "controls")
	if stringAt(controls, "placement") == "header" {
		parts += controlsHTML(object(controls))
	}
	if parts == "" {
		return ""
	}
	styles := []string{}
	if st := stringAt(header, "style"); st != "" {
		styles = append(styles, st)
	}
	if read(vm, "sticky") == true {
		styles = append(styles, "--crudui-sticky-depth: "+stringAt(vm, "stickyDepth"))
	}
	return element("div", NewObject("class", classes("crudui-node__header", stringAt(header, "className")), "style", strings.Join(styles, "; ")), parts)
}
func bodyHTML(vm *Object) string {
	body := read(vm, "body")
	a := NewObject("class", classes("crudui-node__body", stringAt(body, "className")))
	for _, k := range []string{"style", "id"} {
		if has(body, k) {
			a.Set(k, stringAt(body, k))
		}
	}
	inner := ""
	if box := object(read(vm, "checkbox")); box != nil {
		input := NewObject("class", stringAt(box, "className"), "id", stringAt(box, "id"), "name", stringAt(box, "name"), "type", "checkbox", "value", "1")
		if read(box, "checked") == true {
			input.Set("checked", true)
		}
		inner = inputHTML(input, false) + element("label", NewObject("for", stringAt(box, "id")), escape(stringAt(box, "caption")))
	} else if w := object(read(vm, "widget")); w != nil {
		inner = renderWidget(w)
	} else {
		inner = nodesHTML(objectList(read(vm, "children")))
	}
	return "<div" + attrs(a, false, false) + flag("hidden", read(vm, "collapsible") == true && read(vm, "expanded") != true) + ">" + inner + "</div>"
}

// nodeHTML renders one node of the recursive form grammar with its header, body and footer slots.
func nodeHTML(vm *Object) string {
	kind := stringAt(vm, "kind")
	sticky := ""
	if read(vm, "sticky") == true {
		sticky = "crudui-node--sticky"
	}
	a := NewObject("class", classes("crudui-node", "crudui-node--"+kind, sticky, stringAt(vm, "className")))
	if has(vm, "style") {
		a.Set("style", stringAt(vm, "style"))
	}
	if kind != "row" && kind != "lang-item" && has(vm, "path") {
		a.Set("data-field-path", stringAt(vm, "path"))
	}
	if has(vm, "key") {
		a.Set("data-crudui-row-key", stringAt(vm, "key"))
	}
	if has(vm, "lang") {
		a.Set("data-lang", stringAt(vm, "lang"))
	}
	footer := ""
	if controls := read(vm, "controls"); stringAt(controls, "placement") == "footer" {
		footer = element("div", NewObject("class", "crudui-node__footer"), controlsHTML(object(controls)))
	}
	return "<div" + attrs(a, false, false) + flag("hidden", read(vm, "hidden") == true) + ">" + headerHTML(vm) + bodyHTML(vm) + footer + "</div>"
}
func nodesHTML(nodes []*Object) string {
	var out strings.Builder
	for _, n := range nodes {
		out.WriteString(nodeHTML(n))
	}
	return out.String()
}

// RenderForm renders one instance without changing its template or data.
func RenderForm(form *Form) (string, error) {
	if form == nil {
		return "", fmt.Errorf("Form instance is required")
	}
	return `<div class="crudui-form"><div class="crudui-form__body">` + nodesHTML(form.fields) + `</div></div>`, nil
}

var javascriptProtocolRE = regexp.MustCompile(`(?i)^[\x00-\x1f ]*j[\r\n\t]*a[\r\n\t]*v[\r\n\t]*a[\r\n\t]*s[\r\n\t]*c[\r\n\t]*r[\r\n\t]*i[\r\n\t]*p[\r\n\t]*t[\r\n\t]*:`)

func sanitizeURL(value string) string {
	if javascriptProtocolRE.MatchString(value) {
		return "javascript:throw new Error('React has blocked a javascript: URL as a security precaution.')"
	}
	return value
}

func textContent(value string, raw bool) string {
	if raw {
		return escapeText(value)
	}
	return escape(value)
}

func ordinaryAttribute(name string) string {
	switch name {
	case "autocomplete":
		return "autoComplete"
	case "readonly":
		return "readOnly"
	case "maxlength":
		return "maxLength"
	case "minlength":
		return "minLength"
	case "colspan":
		return "colSpan"
	case "rowspan":
		return "rowSpan"
	default:
		return name
	}
}

File: packages/generator-rust/src/render.rs
use serde_json::{json, Map, Value};

use crate::util::{js_string, scalar, style};
use crate::{Form, FormResult};

pub(crate) fn escape(value: &str) -> String {
    value
        .replace('&', "&amp;")
        .replace('<', "&lt;")
        .replace('>', "&gt;")
        .replace('"', "&quot;")
        .replace('\'', "&#x27;")
}

pub(crate) fn raw_text(value: &str) -> String {
    value
        .replace('&', "&amp;")
        .replace('<', "&lt;")
        .replace('>', "&gt;")
}

pub(crate) fn raw_attribute(value: &str) -> String {
    value
        .replace('&', "&amp;")
        .replace('"', "&quot;")
        .replace('<', "&lt;")
}

fn attributes(attrs: &Value, raw: bool) -> String {
    let Some(attrs) = attrs.as_object() else {
        return String::new();
    };
    attrs
        .iter()
        .map(|(key, value)| {
            let value = scalar(Some(value));
            let value = if raw {
                raw_attribute(&value)
            } else {
                escape(&value)
            };
            let key = if raw {
                key.as_str()
            } else {
                match key.as_str() {
                    "autocomplete" => "autoComplete",
                    "readonly" => "readOnly",
                    "maxlength" => "maxLength",
                    "minlength" => "minLength",
                    "colspan" => "colSpan",
                    "rowspan" => "rowSpan",
                    _ => key.as_str(),
                }
            };
            format!(" {key}=\"{value}\"")
        })
        .collect()
}

fn sanitize_url(url: &str) -> String {
    let pattern=regex::RegexBuilder::new(r"^[\x00-\x1f ]*j[\r\n\t]*a[\r\n\t]*v[\r\n\t]*a[\r\n\t]*s[\r\n\t]*c[\r\n\t]*r[\r\n\t]*i[\r\n\t]*p[\r\n\t]*t[\r\n\t]*:")
        .case_insensitive(true).unicode(false).build().expect("URL scheme pattern");
    if pattern.is_match(url) {
        "javascript:throw new Error('React has blocked a javascript: URL as a security precaution.')".into()
    } else {
        url.into()
    }
}

pub(crate) fn element(tag: &str, attrs: &Value, content: &str) -> String {
    let mut attrs = attrs.as_object().cloned().unwrap_or_default();
    if let Some(value) = attrs.get("style").and_then(Value::as_str) {
        let value = crate::css::render_style(value);
        if value.is_empty() {
            attrs.shift_remove("style");
        } else {
            attrs.insert("style".into(), value.into());
        }
    }
    for key in ["href", "src"] {
        if let Some(value) = attrs.get(key).and_then(Value::as_str) {
            if key == "src" && value.is_empty() {
                attrs.shift_remove(key);
            } else {
                attrs.insert(key.into(), sanitize_url(value).into());
            }
        }
    }
    for key in ["readonly", "disabled", "required", "multiple", "autofocus"] {
        if attrs.contains_key(key) {
            attrs.insert(key.into(), "".into());
        }
    }
    if tag == "input" {
        for key in ["name", "checked", "value"] {
            if let Some(value) = attrs.shift_remove(key) {
                attrs.insert(key.into(), value);
            }
        }
    } else if matches!(tag, "select" | "textarea") {
        attrs.shift_remove("value");
    }
    let attributes = attributes(&attrs.into(), false);
    if matches!(tag, "input" | "img" | "br" | "hr" | "link") {
        format!("<{tag}{attributes}/>")
    } else {
        let leading_newline = if tag == "textarea" && content.starts_with('\n') {
            "\n"
        } else {
            ""
        };
        format!("<{tag}{attributes}>{leading_newline}{content}</{tag}>")
    }
}

pub(crate) fn raw_element(tag: &str, attrs: &Value, content: &str) -> String {
    let attributes = attributes(attrs, true);
    if matches!(tag, "input" | "img" | "br" | "hr" | "link") {
        format!("<{tag}{attributes}>")
    } else {
        format!("<{tag}{attributes}>{content}</{tag}>")
    }
}

fn has_events(attrs: &Value) -> bool {
    attrs.as_object().into_iter().flatten().any(|(key, _)| {
        key.strip_prefix("on")
            .and_then(|rest| rest.bytes().next())
            .is_some_and(|byte| byte.is_ascii_lowercase())
    })
}

fn str_at<'a>(value: &'a Value, key: &str) -> &'a str {
    value[key].as_str().unwrap_or("")
}

pub(crate) fn appearance_attrs(class: String, inline: &str) -> Value {
    let mut attrs = Map::new();
    if !class.is_empty() {
        attrs.insert("class".into(), class.into());
    }
    let inline = style(inline);
    if !inline.is_empty() {
        attrs.insert("style".into(), inline.into());
    }
    attrs.into()
}

fn affix(affix: &Value, raw: bool) -> String {
    if !affix.is_object() {
        return String::new();
    }
    let element = if raw { raw_element } else { element };
    let text = if raw { raw_text } else { escape };
    element(
        "span",
        &appearance_attrs(str_at(affix, "class").into(), str_at(affix, "style")),
        &text(str_at(affix, "text")),
    )
}

fn control(widget: &Value, search: bool) -> String {
    let tag = widget["tag"].as_str().unwrap_or("input");
    let raw = search || has_events(&widget["attrs"]);
    let text = if raw { raw_text } else { escape };
    let render = if raw { raw_element } else { element };
    let content = if tag == "select" {
        widget["options"]
            .as_array()
            .into_iter()
            .flatten()
            .map(|option| {
                let mut attrs = json!({"value":option["value"]});
                if option["selected"] == true {
                    attrs["selected"] = if search { "selected" } else { "" }.into();
                }
                render("option", &attrs, &text(str_at(option, "label")))
            })
            .collect()
    } else {
        text(str_at(widget, "text"))
    };
    let mut attrs = widget["attrs"].clone();
    if !raw {
        if let Some(style) = attrs
            .as_object_mut()
            .and_then(|attrs| attrs.shift_remove("style"))
        {
            attrs["style"] = style;
        }
    }
    render(tag, &attrs, &content)
}

fn script(text: &str) -> String {
    element("script", &json!({"nonce":""}), text)
}

fn widget(model: &Value) -> String {
    if model["unsupported"] == true {
        return element(
            "div",
            &json!({"class":"crudui-widget crudui-widget--unsupported","data-unsupported-type":model["type"]}),
            "",
        );
    }
    match str_at(model, "layout") {
        "widget" => element(
            "div",
            &json!({"class":"crudui-widget"}),
            &(affix(&model["prepend"], has_events(&model["attrs"]))
                + &control(model, false)
                + &affix(&model["append"], has_events(&model["attrs"]))),
        ),
        "bare" => control(model, false),
        "host-script" => control(model, false) + &script(str_at(model, "script")),
        "choices" => {
            let radio = model["kind"] == "choice";
            let content = model["options"]
                .as_array()
                .into_iter()
                .flatten()
                .map(|option| {
                    let mut attrs = model["extra"]["input"]
                        .as_object()
                        .cloned()
                        .unwrap_or_default();
                    attrs.insert(
                        "type".into(),
                        if radio { "radio" } else { "checkbox" }.into(),
                    );
                    attrs.insert("value".into(), option["value"].clone());
                    attrs.insert("autocomplete".into(), "off".into());
                    attrs.insert("class".into(), "valid-target crudui-choices__input".into());
                    if option["id"].is_string() {
                        attrs.insert("id".into(), option["id"].clone());
                    }
                    if radio {
                        attrs.insert(
                            "data-is-default".into(),
                            if option["isDefault"] == true { "1" } else { "" }.into(),
                        );
                    }
                    if option["selected"] == true {
                        attrs.insert("checked".into(), "".into());
                    }
                    let label = json!({"for":option["id"],"class":model["itemLabelClass"]});
                    let raw = has_events(&model["extra"]["input"]);
                    let input = if raw { raw_element } else { element };
                    let text = if raw { raw_text } else { escape };
                    input("input", &attrs.into(), "")
                        + &input(
                            "label",
                            &label,
                            &input("span", &json!({}), &text(str_at(option, "label"))),
                        )
                })
                .collect::<String>();
            element("div", &model["attrs"], &content)
        }
        "file" => {
            let raw = has_events(&model["extra"]["file"]);
            let input = if raw { raw_element } else { element };
            let mut content = affix(&model["prepend"], raw);
            if model["extra"]["display"].is_object() {
                let display = if raw {
                    json!({"class":model["extra"]["display"]["class"],"readonly":"","type":"text","value":""})
                } else {
                    model["extra"]["display"].clone()
                };
                content += &input("input", &display, "");
            }
            content += &input("input", &model["extra"]["file"], "");
            if model["extra"]["display"].is_object() {
                content += &element(
                    "button",
                    &json!({"class":"crudui-widget__button","type":"button"}),
                    "&nbsp;",
                );
            }
            element("div", &json!({"class":"crudui-widget"}), &content)
        }
        "display" => element("div", &model["attrs"], str_at(model, "rawHtml")),
        "search" => {
            let mut content = String::new();
            if !str_at(model, "styleChrome").is_empty() {
                content += &element("style", &json!({"nonce":""}), str_at(model, "styleChrome"));
            }
            content += &script(str_at(model, "script"));
            content
                + &element(
                    "div",
                    &json!({"class":"crudui-widget crudui-widget--search"}),
                    &(affix(&model["prepend"], true)
                        + &control(model, true)
                        + &affix(&model["append"], true)),
                )
        }
        "button" => {
            script(str_at(model, "script"))
                + &element("input", &model["extra"]["hidden"], "")
                + &element("input", &model["attrs"], "")
        }
        _ => String::new(),
    }
}

fn classes(parts: &[&str]) -> String {
    parts
        .iter()
        .filter(|part| !part.is_empty())
        .copied()
        .collect::<Vec<_>>()
        .join(" ")
}

/// A present model part as text: strings verbatim, other values as JavaScript strings.
fn part_text(value: &Value) -> String {
    escape(&js_string(value))
}

/// One control group of `data-crudui-action` buttons.
fn controls(controls: &Value) -> String {
    let buttons = controls["actions"]
        .as_array()
        .into_iter()
        .flatten()
        .map(|action| {
            let mut attrs = json!({"type":"button","class":"crudui-action","data-crudui-action":action["name"],"aria-label":action["label"]});
            if action["disabled"] == true {
                attrs["disabled"] = "".into();
            }
            element("button", &attrs, "")
        })
        .collect::<String>();
    element(
        "div",
        &json!({"class":"crudui-controls","role":"group","aria-label":controls["label"]}),
        &buttons,
    )
}

fn header(node: &Value) -> String {
    let header = &node["header"];
    let mut parts = String::new();
    if node["collapsible"] == true {
        let mut attrs = json!({"type":"button","class":"crudui-action","data-crudui-action":"toggle-row",
            "aria-expanded": (node["expanded"] == true).to_string()});
        if let Some(id) = node["body"]["id"].as_str() {
            attrs["aria-controls"] = id.into();
        }
        if let Some(label) = node["toggleLabel"].as_str() {
            attrs["aria-label"] = label.into();
        }
        parts += &element("button", &attrs, "");
    }
    if let Some(label) = header.get("label") {
        parts += &match header.get("labelFor").and_then(Value::as_str) {
            Some(target) if !target.is_empty() => element(
                "label",
                &json!({"class":"crudui-node__label","for":target}),
                &part_text(label),
            ),
            _ => element(
                "span",
                &json!({"class":"crudui-node__label"}),
                &part_text(label),
            ),
        };
    }
    if let Some(description) = header.get("description") {
        parts += &element(
            "p",
            &json!({"class":"crudui-node__description"}),
            &part_text(description),
        );
    }
    for key in ["number", "title"] {
        if let Some(value) = header.get(key) {
            parts += &element(
                "span",
                &json!({"class":format!("crudui-node__{key}")}),
                &part_text(value),
            );
        }
    }
    if let Some(summary) = header.get("summary") {
        let mut attrs = json!({"class":"crudui-node__summary"});
        if node["expanded"] == true {
            attrs["hidden"] = "".into();
        }
        parts += &element("span", &attrs, &part_text(summary));
    }
    if let Some(count) = header.get("count") {
        parts += &element(
            "span",
            &json!({"class":"crudui-node__count"}),
            &part_text(count),
        );
    }
    if node["controls"]["placement"] == "header" {
        parts += &controls(&node["controls"]);
    }
    if parts.is_empty() {
        return String::new();
    }
    element(
        "div",
        &json!({"class":classes(&["crudui-node__header", str_at(header, "className")]),"style":str_at(header, "style")}),
        &parts,
    )
}

fn body(node: &Value) -> String {
    let body = &node["body"];
    let mut attrs = json!({"class":classes(&["crudui-node__body", str_at(body, "className")]),"style":str_at(body, "style")});
    if let Some(id) = body["id"].as_str().filter(|id| !id.is_empty()) {
        attrs["id"] = id.into();
    }
    if node["collapsible"] == true && node["expanded"] != true {
        attrs["hidden"] = "".into();
    }
    let content = if node["checkbox"].is_object() {
        let checkbox = &node["checkbox"];
        let mut input = json!({"class":checkbox["className"],"id":checkbox["id"],"name":checkbox["name"],"type":"checkbox","value":"1"});
        if checkbox["checked"] == true {
            input["checked"] = "".into();
        }
        element("input", &input, "")
            + &element(
                "label",
                &json!({"for":checkbox["id"]}),
                &escape(str_at(checkbox, "caption")),
            )
    } else if node.get("widget").is_some() {
        widget(&node["widget"])
    } else {
        nodes(&node["children"])
    };
    element("div", &attrs, &content)
}

/// Render one node of the recursive form grammar.
fn node(node: &Value) -> String {
    let kind = str_at(node, "kind");
    let modifier = format!("crudui-node--{kind}");
    let sticky = if node["sticky"] == true {
        "crudui-node--sticky"
    } else {
        ""
    };
    // A sticky row carries its depth on the root; the stylesheet derives its sticky line from it.
    let depth = if node["sticky"] == true {
        format!(
            "--crudui-sticky-depth: {}",
            node["stickyDepth"].as_u64().unwrap_or(0)
        )
    } else {
        String::new()
    };
    let style = [str_at(node, "style"), depth.as_str()]
        .into_iter()
        .filter(|s| !s.is_empty())
        .collect::<Vec<_>>()
        .join("; ");
    let mut attrs = json!({"class":classes(&["crudui-node", &modifier, sticky, str_at(node, "className")]),"style":style});
    if kind != "row" && kind != "lang-item" {
        if let Some(path) = node.get("path") {
            attrs["data-field-path"] = js_string(path).into();
        }
    }
    if let Some(key) = node.get("key") {
        attrs["data-crudui-row-key"] = js_string(key).into();
    }
    if let Some(lang) = node.get("lang") {
        attrs["data-lang"] = js_string(lang).into();
    }
    if node["hidden"] == true {
        attrs["hidden"] = "".into();
    }
    let footer = if node["controls"]["placement"] == "footer" {
        element(
            "div",
            &json!({"class":"crudui-node__footer"}),
            &controls(&node["controls"]),
        )
    } else {
        String::new()
    };
    element("div", &attrs, &(header(node) + &body(node) + &footer))
}

fn nodes(nodes: &Value) -> String {
    nodes.as_array().into_iter().flatten().map(node).collect()
}

pub(crate) fn render_fields(fields: &[Value]) -> String {
    element(
        "div",
        &json!({"class":"crudui-form"}),
        &element(
            "div",
            &json!({"class":"crudui-form__body"}),
            &fields.iter().map(node).collect::<String>(),
        ),
    )
}

/// Render fields and the form buttons for a record as form HTML.
pub(crate) fn render_form_html(
    fields: &[Value],
    template: &crate::FormTemplate,
    data: &Value,
    language: &str,
) -> FormResult<String> {
    let messages = crate::messages::form_messages(language)?;
    let body = render_fields(fields);
    let footer = element(
        "div",
        &json!({"class":"crudui-form__footer"}),
        &element(
            "div",
            &json!({"class":"crudui-controls","role":"group","aria-label":messages.form_actions}),
            &crate::buttons::form_buttons_html(&template.buttons, data, language, messages),
        ),
    );
    Ok(format!(
        "{}{footer}</div>",
        body.strip_suffix("</div>")
            .expect("form markup ends with its closing tag")
    ))
}

/// Render an instance as HTML without executing scripts or browser operations.
pub fn render_form(form: &Form) -> FormResult<String> {
    render_form_html(
        form.fields(),
        form.template(),
        &form.get_data(),
        form.language(),
    )
}
